internal/pkg/controllers/binding/binding.go

```diff
@@ -0,0 +1,260 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package binding
+
+import (
+	"context"
+	"strings"
+	"time"
+
+	"github.com/crossplane/crossplane-runtime/pkg/event"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"sigs.k8s.io/security-profiles-operator/api/profilebinding/v1alpha1"
+)
+
+const (
+	reconcileTimeout = 1 * time.Minute
+
+	errGetBinding = "cannot get binding"
+
+	reasonCannotUpdateProfileBinding       event.Reason = "CannotUpdateProfileBinding"
+	reasonCannotUpdateProfileBindingStatus event.Reason = "CannotUpdateProfileBindingStatus"
+	reasonCannotRetrieveProfileBinding     event.Reason = "CannotRetrieveProfileBinding"
+
+	profileRefKey = ".metadata.profileRef"
+	boundPodsKey  = ".metadata.boundPods"
+
+	finalizer = "active-workload-lock"
+)
+
+// Setup adds a controller that reconciles profile bindings.
+func Setup(ctx context.Context, mgr ctrl.Manager, l logr.Logger) error {
+	// Index ProfileBindings by the Profile they reference
+	if err := mgr.GetFieldIndexer().IndexField(
+		ctx, &v1alpha1.ProfileBinding{}, profileRefKey, func(rawObj runtime.Object) []string {
+			pb, ok := rawObj.(*v1alpha1.ProfileBinding)
+			if !ok {
+				return []string{}
+			}
+			return []string{pb.GetNamespace() + "/" + pb.Spec.Profile}
+		}); err != nil {
+		return errors.Wrap(err, "creating profile index on profilebinding")
+	}
+	// Index ProfileBindings by their active workloads
+	if err := mgr.GetFieldIndexer().IndexField(
+		ctx, &v1alpha1.ProfileBinding{}, boundPodsKey, func(rawObj runtime.Object) []string {
+			pb, ok := rawObj.(*v1alpha1.ProfileBinding)
+			if !ok {
+				return []string{}
+			}
+			return pb.Status.ActiveWorkloads
+		}); err != nil {
+		return errors.Wrap(err, "creating pod index on profilebinding")
+	}
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named("pods").
+		For(&corev1.Pod{}).
+		WithEventFilter(resource.NewPredicates(hasSeccompProfile)).
+		Complete(&Reconciler{
+			client: mgr.GetClient(),
+			log:    l,
+			record: event.NewAPIRecorder(mgr.GetEventRecorderFor("profilebinding")),
+		})
+}
+
+type Reconciler struct {
+	client client.Client
+	log    logr.Logger
+	record event.Recorder
+}
+
+func (r *Reconciler) Reconcile(req reconcile.Request) (reconcile.Result, error) {
```

Comment:
if we use kubebuilder more, we could set the rbac permissions dynamically here: https://book.kubebuilder.io/reference/markers/rbac.html this can ensure that we always have up-to-date permissions.